### PR TITLE
Bump Julia to LLVM 16

### PR DESCRIPTION
**Pull request by vchuravy**
Current Status:

- [x] Self test of Julia-GCChecker :: `MissingRoots.c`
- [x] ASAN: `error: <unknown>:0: Cannot represent a difference across sections`
- [x] x86_64-apple-darwin fails in ranges 
- [x] llvmpasses

### CODE REVIEW

**Comment by vchuravy** [comment]
```
/home/vchuravy/src/julia2/src/jitlayers.cpp: In lambda function:
/home/vchuravy/src/julia2/src/jitlayers.cpp:1675:19: error: 'class llvm::LLVMContext' has no member named 'hasSetOpaquePointersValue'; did you mean 'setOpaquePointers'?
 1675 |         if (!ctx->hasSetOpaquePointersValue())
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                   setOpaquePointers
make[1]: *** [Makefile:250: jitlayers.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:97: julia-src-release] Error 2
```

cc: @maleadt

**Comment by vtjnash** [comment]
LLVM deleted this functionality, claimed NFC for the change
```
commit 4756f9ae0aa6c2a0b4231d93d5166c055f303603
Author: Nikita Popov <npopov@redhat.com>
Date:   Wed Jan 18 11:04:12 2023 +0100

    [LLVMContext] Remove hasSetOpaquePointersValue() API (NFC)
```

**Comment by vtjnash** [comment]
You now need to check `OpaquePointersCL.getNumOccurrences()` to get that info (like we do during initialization--you could probably memoize that check into a global variable in our library as well to make it easy on yourself)

**Comment by gbaraldi** [comment]
Checking the command line isn't enough, because we want to respect wheter GPUCompiler added a value or not.

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(ALL, vs="master")`

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(vs=":master")`

**Comment by gbaraldi** [comment]
@nanosoldier runbenchmarks(ALL,vs=":master")

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks()`

**Comment by gbaraldi** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1826199945) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2704710_vs_187e8c2/report.html) is available.

**Comment by vchuravy** [comment]
- i686-windows:
```
cd /c/workdir/base && /c/workdir/usr/bin/julia.exe -C "pentium4" --heap-size-hint=1000M --output-ji `cygpath -w /c/workdir/usr/lib/julia/corecompiler.ji`.tmp --startup-file=no --warn-overwrite=yes -g0 -O0 compiler/compiler.jl
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x39e1b6f -- _ZN4llvm9Intrinsic28getIntrinsicInfoTableEntriesEjRNS_15SmallVectorImplINS0_13IITDescriptorEEE at C:\workdir\usr\bin\libLLVM-16jl.dll (unknown line)
in expression starting at boot.jl:263
_ZN4llvm9Intrinsic28getIntrinsicInfoTableEntriesEjRNS_15SmallVectorImplINS0_13IITDescriptorEEE at C:\workdir\usr\bin\libLLVM-16jl.dll (unknown line)
Allocations: 0 (Pool: 0; Big: 0); GC: 0
make[1]: *** [sysimage.mk:61: /c/workdir/usr/lib/julia/corecompiler.ji] Error 1
make[1]: Leaving directory '/c/workdir'
```
In the past I had success using a full asan build of LLVM to reproduce these on Linux

**Comment by gbaraldi** [comment]
I tried a full ASAN + debug + julia debug and got nothing on linux
>  JULIA tmp/test-asan/asan/usr/lib/julia/corecompiler.ji
> Core.Compiler ──── 1092.12 seconds

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1830134920) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/80d7ee4_vs_68b4587/report.md).

**Comment by gbaraldi** [comment]
@nanosoldier runbenchmarks(!"scalar", vs=":master")

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1850979950) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/152a74d_vs_b71523e/report.md).

**Comment by gbaraldi** [comment]
Nanosoldier ran on the wrong commit :(

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1852599721) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/152a74d_vs_eba10dd/report.md).

**Comment by vchuravy** [comment]
I think this is the best error for the Win32 issue yet

```
Exception: EXCEPTION_ACCESS_VIOLATION at 0x3d71b6f -- _ZN4llvm9Intrinsic28getIntrinsicInfoTableEntriesEjRNS_15SmallVectorImplINS0_13IITDescriptorEEE at C:\workdir\usr\bin\libLLVM-16jl.dll (unknown line)
in expression starting at boot.jl:263
_ZN4llvm9Intrinsic28getIntrinsicInfoTableEntriesEjRNS_15SmallVectorImplINS0_13IITDescriptorEEE at C:\workdir\usr\bin\libLLVM-16jl.dll (unknown line)
Allocations: 0 (Pool: 0; Big: 0); GC: 0
```

**Comment by vchuravy** [comment]
Current Status:

1. Self test of Julia-GCChecker :: `MissingRoots.c`
2. ASAN: `error: <unknown>:0: Cannot represent a difference across sections`
3. x86_64-apple-darwin fails in ranges 
4. llvmpasses

**Comment by gbaraldi** [comment]
So ranges seems to be very small floating point differences, it might be interesting to look why they are there, but I think it's a question of making the test more robust.

**Comment by gbaraldi** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1860736979) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3b719a0_vs_2ab4105/report.md).

File: src/jitlayers.cpp [hunk 1]
```diff
@@ -1665,6 +1665,26 @@ void fixupTM(TargetMachine &TM){
...
+}
+
+extern "C" JL_DLLEXPORT_CODEGEN
+int JuliaUsesOpaquePtrs() {
+#ifndef JL_LLVM_OPAQUE_POINTERS
+    return 0;
+#else
+    return 1;
+#endif
+}
```
Comment (maleadt) [inline src/jitlayers.cpp:1686]: Unused? Or do you have an external use for this?

**Comment by gbaraldi** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51720#issuecomment-1871992545) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b61062b_vs_66e9410/report.html) is available.

**Comment by gbaraldi** [comment]
After lots of annoyances and setbacks, I think this is ready for review. PkgEval looks good. the AllocCheck segfault is weird but I'm taking a look at it.

File: src/aotcompile.cpp [hunk 2]
```diff
@@ -1051,6 +1051,20 @@ static AOTOutputs add_output_impl(Module &M, TargetMachine &SourceTM, ShardTimer
             // functions. We do so after optimization to avoid cloning these functions.
 
             // Float16 conversion routines
+#if defined(_CPU_X86_64_) && defined(_OS_DARWIN_) && JL_LLVM_VERSION >= 160000
```
Comment (pchintalapudi) [inline src/aotcompile.cpp:1054]: Can we make this dependent on the target triple? Just to preserve the fiction that codegen is cross-compile capable.

File: src/cgutils.cpp [hunk 3]
```diff
@@ -911,7 +911,7 @@ static bool is_tupletype_homogeneous(jl_svec_t *t, bool allow_va = false)
 }
 
 static bool for_each_uniontype_small(
-        std::function<void(unsigned, jl_datatype_t*)> f,
+        const std::function<void(unsigned, jl_datatype_t*)> &f,
```
Comment (pchintalapudi) [inline src/cgutils.cpp:914]: Could use an llvm::function_ref here.

File: src/jitlayers.cpp [hunk 4]
```diff
@@ -1665,6 +1665,17 @@ void fixupTM(TargetMachine &TM){
     }
 }
 
+extern int jl_opaque_ptrs_set;
+void SetOpaquePointer(LLVMContext &ctx) {
+    if (jl_opaque_ptrs_set)
+        return;
+#ifndef JL_LLVM_OPAQUE_POINTERS
```
Comment (pchintalapudi) [inline src/jitlayers.cpp:1672]: Is anyone actually using this build flag, given we have a runtime option for setting it as well? I'm in favor of just deleting.

**Review by pchintalapudi — APPROVED** [review]

**Comment by vchuravy** [comment]
Congratulations on getting this over the line!

**Pull request merged**